File: 42-TrappingRainWater/42-TrappingRainWater.java
// Last updated: 8/19/2026, 10:41:36 AM
class Solution {
    public int trap(int[] a) {
        int n=a.length;
        int left[] = new int[n];
        int right[] = new int[n];
        int lmax=0;
        int rmax=0;
        for(int i=0;i<n;i++)
        {
        if(lmax<a[i])
        {
            lmax=a[i];
            left[i]=lmax;
        }
        else
        left[i]=lmax;
        }
         for(int i=n-1;i>=0;i--)
         {
            if(rmax<a[i])
            {
                rmax=a[i];
                right[i]=rmax;
            }
            else
            right[i]=rmax;
         }
         int total=0;
         for(int i=0;i<n;i++)
         {
            int min=left[i]<right[i]?left[i]:right[i];
            total=total+(min-a[i]);
         }
         return total;
    }
}

File: 42-TrappingRainWater/42-TrappingRainWater.py
# Last updated: 8/19/2026, 10:41:43 AM
class Solution:
    def trap(self, height: List[int]) -> int:
        leftwall = rightwall = 0
        n = len(height)
        left = [0] *n
        right = [0] * n
        for i in range(n):
            j=-i-1
            left[i]=leftwall
            right[j]=rightwall
            leftwall=max(leftwall,height[i])
            rightwall=max(rightwall,height[j])
        total=0
        for k in range(n):
            minimum = min(left[k], right[k])
            total += max(0,minimum-height[k])
        return total

        